kimchi/src/snarky/traits.rs

```diff
@@ -0,0 +1,136 @@
+//! Traits.
+
+use ark_ff::PrimeField;
+
+use super::{
+    checked_runner::{RunState, WitnessGeneration},
+    cvar::CVar,
+    prelude::TypeCreation,
+};
+
+/// A snarky type is a type that can be used in a circuit.
+/// It references an equivalent "out-of-circuit" type that one can use outside of the circuit.
+/// (For example, to construct private or public inputs, or a public output, to the circuit.)
+pub trait SnarkyType<F>: Sized
+where
+    F: PrimeField,
+{
+    /// Some additional information. (TODO: add more documentation.)
+    type Auxiliary;
+
+    /// The equivalent "out-of-circuit" type.
+    /// For example, the [super::boolean::Boolean] snarky type has an out-of-circuit type of [bool].
+    type OutOfCircuit;
+
+    /// The number of field elements that this type takes.
+    const SIZE_IN_FIELD_ELEMENTS: usize;
+
+    /// Returns the circuit variables (and auxiliary data) behind this type.
+    fn to_cvars(&self) -> (Vec<CVar<F>>, Self::Auxiliary);
+
+    /// Creates a new instance of this type from the given circuit variables (And some auxiliary data).
+    fn from_cvars_unsafe(cvars: Vec<CVar<F>>, aux: Self::Auxiliary) -> Self;
+
+    /// Checks that the circuit variables behind this type are valid.
+    /// For some definition of valid.
+    /// For example, a Boolean snarky type would check that the field element representing it is either 0 or 1.
+    /// The function does this by adding constraints to your constraint system.
+    fn check(&self, cs: &mut RunState<F>);
+
+    /// Deserializes this type into the out-of-circuit type (along with some auxiliary data).
+    fn deserialize(&self) -> (Self::OutOfCircuit, Self::Auxiliary);
+
+    /// Serializes the out-of-circuit type (along with some auxiliary data) into this type.
+    fn serialize(out_of_circuit: Self::OutOfCircuit, aux: Self::Auxiliary) -> Self;
+
+    /// Returns some auxiliary data (TODO: more doc).
```

Comment:
The 'default' value of `Self::Auxiliary`. This is passed to `of_cvars_unsafe` when we are not generating a witness, since we have no candidate value to get the `Auxiliary` from.

Note that we use an explicit value here rather than `Auxiliary: Default`, since the default value for the type may not match the default value we actually want to pass!

Reply:
oh I see, wouldn't it make sense to have two sets of functions, functions that take an auxiliary as param, and functions that don't?